sklearn/preprocessing/data.py

```diff
@@ -2440,3 +2442,147 @@ def quantile_transform(X, axis=0, n_quantiles=1000,
+    ... # doctest: +ELLIPSIS
+    CategoricalEncoder(categories='auto', dtype=<... 'numpy.float64'>,
+              handle_unknown='error', sparse=True)
+    >>> enc.transform([[0, 1, 1]]).toarray()
+    array([[ 1.,  0.,  0.,  1.,  0.,  0.,  1.,  0.,  0.]])
+
+    See also
+    --------
+    sklearn.preprocessing.OneHotEncoder : performs a one-hot encoding of
+      integer ordinal features. This transformer assumes that input features
+      take on values in the range [0, max(feature)].
+    sklearn.feature_extraction.DictVectorizer : performs a one-hot encoding of
+      dictionary items (also handles string-valued features).
+    sklearn.feature_extraction.FeatureHasher : performs an approximate one-hot
+      encoding of dictionary items or strings.
+    """
+
+    def __init__(self, categories='auto', dtype=np.float64, sparse=True,
+                 handle_unknown='error'):
+        self.categories = categories
+        self.dtype = dtype
+        self.sparse = sparse
+        self.handle_unknown = handle_unknown
+
+    def fit(self, X, y=None):
+        """Fit the CategoricalEncoder to X.
+
+        Parameters
+        ----------
+        X : array-like, shape [n_samples, n_feature]
+            Input array of type int.
+
+        Returns
+        -------
+        self
+        """
+
+        if self.handle_unknown not in ['error', 'ignore']:
+            template = ("handle_unknown should be either 'error' or "
+                        "'ignore', got %s")
+            raise ValueError(template % self.handle_unknown)
+
+        X = check_array(X, dtype=np.object, accept_sparse='csc', copy=True)
+        n_samples, n_features = X.shape
+
+        self._label_encoders_ = [LabelEncoder() for _ in range(n_features)]
+
+        for i in range(n_features):
+            le = self._label_encoders_[i]
+            Xi = X[:, i]
+            if self.categories == 'auto':
+                le.fit(Xi)
+            else:
+                if not np.all(np.in1d(Xi, self.categories[i])):
+                    if self.handle_unknown == 'error':
+                        diff = np.setdiff1d(Xi, self.categories[i])
+                        msg = 'Unknown feature(s) %s in column %d' % (diff, i)
+                        raise ValueError(msg)
+                le.classes_ = np.array(np.sort(self.categories[i]))
+
+        return self
+
+    def transform(self, X, y=None):
+        """Encode the selected categorical features using the one-hot scheme.
+        """
+        X = check_array(X, accept_sparse='csc', dtype=np.object, copy=True)
+        n_samples, n_features = X.shape
+        X_int = np.zeros_like(X, dtype=np.int)
+        X_mask = np.ones_like(X, dtype=np.bool)
+
+        for i in range(n_features):
+            valid_mask = np.in1d(X[:, i], self._label_encoders_[i].classes_)
+
+            if not np.all(valid_mask):
+                if self.handle_unknown == 'error':
+                    diff = np.setdiff1d(X[:, i],
```

Comment:
It seems to make it faster using this toy example:

```
In [31]: X = np.random.choice(list('abcde'), size=10000)

In [32]: classes = np.array(list('abcd'))

In [33]: valid_mask = np.in1d(X, classes)

In [34]: np.setdiff1d(X, classes)
Out[34]: 
array(['e'], 
      dtype='<U1')

In [35]: np.unique(X[~valid_mask])
Out[35]: 
array(['e'], 
      dtype='<U1')

In [36]: %timeit np.setdiff1d(X, classes)
567 µs ± 2.85 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [37]: %timeit np.unique(X[~valid_mask])
114 µs ± 1.78 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

On bigger X / more categories the difference becomes even larger. So will change this.